northstar/src/runtime/console.rs

```diff
@@ -452,6 +268,197 @@ impl Console {
+        } else {
+            send_event(&event_loop, Request::Message(message)).await?
+        };
+
+    Ok(api::model::Message {
+        id: message_id,
+        payload: api::model::Payload::Response(response),
+    })
+}
+
+/// Sends an `Event::Console` to the event loop and await its response
+async fn send_event(event_loop: &EventTx, request: Request) -> Result<api::model::Response, Error> {
+    trace!("{:?} -> event loop", request);
+    let (reply_tx, reply_rx) = oneshot::channel();
+    let response = event_loop
+        .send(Event::Console(request, reply_tx))
+        .map_err(|_| Error::EventLoopClosed)
+        .and_then(|_| reply_rx.map_err(|_| Error::EventLoopClosed))
+        .await;
+    trace!("{:?} <- event loop", response);
+    response
+}
+
+/// Types of listeners for console connections
+enum Listener {
+    Tcp(TcpListener),
+    Unix(UnixListener),
+}
+
+impl Listener {
+    async fn new(url: &Url) -> std::io::Result<Listener> {
+        let listener = match url.scheme() {
+            "tcp" => {
+                let address = url.socket_addrs(|| Some(4200))?.first().unwrap().to_owned();
+                debug!("Starting console on {}", &address);
+                let listener = TcpListener::bind(&address).await?;
+                debug!("Started console on {}", &address);
+
+                Listener::Tcp(listener)
+            }
+            "unix" => {
+                let path = PathBuf::from(url.path());
+                debug!("Starting console on {}", path.display());
+
+                // TODO this file should not be deleted here
+                if path.exists() {
+                    fs::remove_file(&path).await?
+                }
+
+                let listener = UnixListener::bind(&path)?;
+
+                debug!("Started console on {}", path.display());
+                Listener::Unix(listener)
+            }
+            _ => unreachable!(),
+        };
+        Ok(listener)
+    }
+}
+
+/// Function to handle connections
+///
+/// Generic handling of connections. The first parameter is a function that when called awaits for
+/// a new connection. The connections are represented as a pair of a stream and some client
+/// identifier.
+///
+/// All the connections container stored the tasks corresponding to each active connection. As
+/// these tasks terminate, they are removed from the connections container. Once a stop is issued,
+/// the termination of the remaining connections will be awaited.
+///
+async fn handle_connections<AcceptConnection, Connection, Stream, Client, E>(
+    accept: AcceptConnection,
+    event_tx: EventTx,
+    notification_tx: broadcast::Sender<Notification>,
+    stop: CancellationToken,
+) where
+    AcceptConnection: Fn() -> Connection,
+    Connection: Future<Output = Result<(Stream, Client), E>>,
+    Stream: AsyncWrite + AsyncRead + Unpin + Send + 'static,
+    Client: Into<ClientId>,
+    E: std::fmt::Debug,
+{
+    let mut connections = FuturesUnordered::new();
```

Comment:
didn't know about this `FuturesUnordered`...where did you learn about that?

Reply:
this one is used already in master, @flxo  introduced it in the last update to `console.rs`